src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -225,9 +218,110 @@ private void pre20Upgrade() throws Exception {
+        if (newRoutingNode == null) {
+            // this needs some other handling
+            return indicesToWrite.values();
+        }
+        // iterate over all shards allocated on this node in the new cluster state but only write if ...
+        for (MutableShardRouting shardRouting : newRoutingNode) {
+            IndexMetaData indexMetaData = newState.metaData().index(shardRouting.index());
+            IndexMetaData currentIndexMetaData = currentMetaData == null ? null : currentMetaData.index(indexMetaData.index());
+            String writeReason = null;
+            // ... state persistence was disabled or index was newly created
+            if (currentIndexMetaData == null) {
+                writeReason = "freshly created";
+                // ... new shard is allocated on node (we could optimize here and make sure only written once and not for each shard per index -> do later)
+            } else if (shardsOfThisIndexExistedOnNodeAlready(previousRoutingNode, shardRouting) == false) {
+                writeReason = "newly allocated on node";
+                // ... version changed
+            } else if (indexMetaData.version() != currentIndexMetaData.version()) {
+                writeReason = "version changed from [" + currentIndexMetaData.version() + "] to [" + indexMetaData.version() + "]";
+            }
+            if (writeReason != null) {
+                indicesToWrite.put(shardRouting.index(),
+                        new GatewayMetaState.IndexMetaWriteInfo(indexMetaData, currentIndexMetaData,
+                                writeReason));
+            }
+        }
+        return indicesToWrite.values();
+    }
+
+    public static boolean shardsOfThisIndexExistedOnNodeAlready(RoutingNode previousRoutingNode, MutableShardRouting shardRouting) {
+        return previousRoutingNode.shardsWithState(shardRouting.index(), ShardRoutingState.INITIALIZING, ShardRoutingState.STARTED).size() != 0;
+    }
+
+    /**
+     * Loads the current meta state for each index in the new cluster state and checks if it has to be persisted.
+     * Each index state that is part of the new cluster state will be considered even if this node has no shard of the
+     * index allocated on it. This is only run for master nodes.
+     *
+     * @param state           the cluster state from which we figure out what is new in each index and should potentially be written
+     * @param currentMetaData the current index state in memory.
+     * @return iterable over all indices states that should be written to disk
+     */
+    public static Iterable<GatewayMetaState.IndexMetaWriteInfo> filterStatesOnMaster(ClusterState state, MetaData currentMetaData) {
+        Map<String, GatewayMetaState.IndexMetaWriteInfo> indicesToWrite = new HashMap<>();
+        MetaData newMetaData = state.metaData();
+        // iterate over all indices but only write if ...
+        for (IndexMetaData indexMetaData : newMetaData) {
```

Comment:
I think we can merge this method body with the different one, no? the only difference is which indices we begin with (all of them or just those allocated to this node).